### PR TITLE
Test Buttons On Gui

### PR DESCRIPTION
**Pull request by sithlord48**
+ ui improvments
+ kps test buttons
+ mostly working print

### CODE REVIEW

File: src/mainwindow.cpp [hunk 1]
```diff
@@ -102,31 +113,103 @@ void MainWindow::locateSerialPort()
...
+void MainWindow::bedTempPBClicked()
+{
+    addSLog(QString(tr("Set Bed Temp: %1").arg(QString::number(ui->bedTempSB->value()))));
+    pro->pushCommand(QString("M140 S%1").arg(ui->bedTempSB->value()).toLocal8Bit());
+}
+
+void MainWindow::extTempPBClicked()
+{
+    addSLog(QString(tr("Set Extruder(%1) Temp: %2").arg(ui->extTempSelCB->currentText().at(9), QString::number(ui->extTempSB->value()))));
+    pro->pushCommand(QString("M104 P%1 S%2").arg(ui->extTempSelCB->currentText().at(9)).arg(ui->extTempSB->value()).toLocal8Bit());
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: I am not a huge fan of 80 chars, but please, split this ! It's hard to read it.

File: src/mainwindow.cpp [hunk 2]
```diff
@@ -102,31 +113,103 @@ void MainWindow::locateSerialPort()
...
 
+void MainWindow::fanSpeedPBClicked()
+{
+    addSLog(QString(tr("Set Fan(%1) Speed: %2%").arg(ui->fanSpeedSelCB->currentText().at(4), QString::number(ui->fanSpeedSB->value()))));
+    pro->pushCommand(QString("M106 P%1 S%2").arg(ui->fanSpeedSelCB->currentText().at(4)).arg(ui->fanSpeedSB->value()).toLocal8Bit());
+}
+
+void MainWindow::printPBClicked()
+{
+    QString fileName =QFileDialog::getOpenFileName(this, tr("Select A file to print"), QDir::homePath(), tr("*.gcode"));
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: [space] = [space]

Comment (patrickelectric) [inline src/mainwindow.cpp]: Why .gcode need to use tr ?

File: src/mainwindow.cpp [hunk 3]
```diff
@@ -102,31 +113,103 @@ void MainWindow::locateSerialPort()
...
+    if(fileName.isNull()){
+    addLog(tr("No File Selected"));
+    }
+    else{
+        addLog(QString(tr("print: %1")).arg(fileName));
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Remove QString

File: src/mainwindow.cpp [hunk 4]
```diff
@@ -102,31 +113,103 @@ void MainWindow::locateSerialPort()
...
+
+        while(file.canReadLine()){
+            addLog(tr("Reading File.."));
+            cline=file.readLine();
+            addLog(QString(tr("NextLine: %1")).arg(cline));
+            if(cline.contains(QChar(';'))){
+                cline.resize(cline.indexOf(QChar(';')));
+            }
+            cline.append("\n\r");
+            addLog(QString("Next Command: %1").arg(cline));
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: tr

File: src/mainwindow.cpp [hunk 5]
```diff
@@ -102,31 +113,114 @@ void MainWindow::locateSerialPort()
...
+    else{
+        addLog(tr("print: %1").arg(fileName));
+        QFile file(fileName);
+        file.open(QFile::ReadOnly);
+        QTextStream gcodestream(&file);
+        QString cline;
+
+        addLog(tr("Reading File.."));
+        while(!gcodestream.atEnd()){
+            cline=gcodestream.readLine();
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: [space] = [space]

File: src/mainwindow.cpp [hunk 6]
```diff
@@ -102,31 +113,114 @@ void MainWindow::locateSerialPort()
...
+            addLog(tr("NextLine: %1").arg(cline));
+            if(cline.contains(QChar(';'))){
+                cline.resize(cline.indexOf(QChar(';')));
+            }
+            if(!cline.isEmpty()){
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Before this if, add cline = cline.simplified(); to remove unused things

File: src/mainwindow.cpp [hunk 7]
```diff
@@ -102,31 +113,114 @@ void MainWindow::locateSerialPort()
...
+                addLog(tr("Next Command: %1").arg(cline));
+                pro->pushCommand(cline.toLocal8Bit());
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Add \n\r in cline

File: src/mainwindow.cpp [hunk 8]
```diff
@@ -102,31 +113,114 @@ void MainWindow::locateSerialPort()
...
+            }
+        }
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Add inside the while loop QCoreApplication::processEvents(); to not freeze the interface

Comment (patrickelectric) [inline src/mainwindow.cpp]: processEvents(); isn't the best option. Its necessary to use QEventLoop.
Take a look [kere](https://forum.qt.io/topic/12742/solved-wait-for-timeout-or-some-signal/10).

File: src/mainwindow.cpp (hunk 1 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: Fixed!

File: src/mainwindow.cpp (hunk 2 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: Fixed both

File: src/mainwindow.cpp (hunk 3 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: fixed

File: src/mainwindow.cpp (hunk 4 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: Fixed

File: src/mainwindow.cpp (hunk 5 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: fixed

File: src/mainwindow.h [hunk 9]
```diff
@@ -29,10 +28,73 @@ public slots:
...
+     */
+    void homeZPBClicked();
+
+    /**
+     * @brief Set Bed Temp Clicked
+     */
+    void bedTempPBClicked();
+
+    /**
+     * @brief Set Ext Temp Clicked
```
Comment (patrickelectric) [inline src/mainwindow.h]: Extruder, is it ?

File: src/mainwindow.cpp [hunk 10]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
     }
 
 }
+void MainWindow::connectPBClicked()
+{
+    QString port = ui->serialPortCB->currentText();
+    uint baud = ui->baudRateLE->text().toUInt();
+    pro = new ProtocolLayer(port, baud);
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: its necessary to delete pro if it already exist.
Like connect disconnect and connect again.

File: src/mainwindow.cpp [hunk 11]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+    connect(pro, &ProtocolLayer::receivedCommand, this, &MainWindow::checkReceivedCommand);
+    connect(pro, &ProtocolLayer::pushedCommand, this, &MainWindow::checkPushedCommands);
+    ui->logTE->clear();
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Maybe a clear button ?

File: src/mainwindow.cpp [hunk 12]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+void MainWindow::homeZPBClicked()
+{
+    addSLog(tr("Home Z"));
+    pro->pushCommand(QString("G28 Z0").toLocal8Bit());
+}
 
-    } else if(btn == ui->disconnectPB){
+void MainWindow::bedTempPBClicked()
+{
+    addSLog(QString(tr("Set Bed Temp: %1")
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Remove QString

File: src/mainwindow.cpp [hunk 13]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+        .arg(QString::number(ui->bedTempSB->value()))));
+    pro->pushCommand(QString("M140 S%1")
+        .arg(ui->bedTempSB->value()).toLocal8Bit());
+}
 
-        pro->closeConnection();
-        addLog(tr("Disconnected"));
+void MainWindow::extTempPBClicked()
+{
+    addSLog(QString(tr("Set Extruder(%1) Temp: %2")
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Remove QString

File: src/mainwindow.cpp [hunk 14]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+        .arg(ui->extTempSelCB->currentText().at(9), QString::number(ui->extTempSB->value()))));
+    pro->pushCommand(QString("M104 P%1 S%2")
+        .arg(ui->extTempSelCB->currentText().at(9), QString::number(ui->extTempSB->value()))
+        .toLocal8Bit());
+}
 
-    } else if(btn == ui->sendPB){
+void MainWindow::mvAxisPBClicked()
+{
+    addSLog(QString(tr("Move %1 to %2")
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Remove QString

File: src/mainwindow.cpp [hunk 15]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+        .arg(ui->mvAxisCB->currentText().at(5), QString::number(ui->mvAxisSB->value()))
+        .toLocal8Bit());
+}
 
-        QByteArray comm = ui->commandLE->text().toUpper().toLocal8Bit();
-        pro->pushCommand(comm);
-        ui->commandLE->clear();
+void MainWindow::fanSpeedPBClicked()
+{
+    addSLog(QString(tr("Set Fan(%1) Speed: %2%")
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Remove QString

File: src/mainwindow.cpp [hunk 16]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+        .arg(ui->fanSpeedSelCB->currentText().at(4), QString::number(ui->fanSpeedSB->value()))
+        .toLocal8Bit());
+}
 
+void MainWindow::printPBClicked()
+{
+    QString fileName = QFileDialog::getOpenFileName(this, tr("Select A file to print"), QDir::homePath(), "*.gcode");
+    addLog(tr("file picked: %1").arg(fileName));
+    if(fileName.isNull()){
+    addLog(tr("No File Selected"));
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: Indentation

File: src/mainwindow.cpp [hunk 17]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+        .arg(ui->fanSpeedSelCB->currentText().at(4), QString::number(ui->fanSpeedSB->value()))));
+    pro->pushCommand(QString("M106 P%1 S%2")
+        .arg(ui->fanSpeedSelCB->currentText().at(4), QString::number(ui->fanSpeedSB->value()))
+        .toLocal8Bit());
+}
 
+void MainWindow::printPBClicked()
+{
+    QString fileName = QFileDialog::getOpenFileName(this, tr("Select A file to print"), QDir::homePath(), "*.gcode");
+    addLog(tr("file picked: %1").arg(fileName));
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: File

File: src/mainwindow.cpp [hunk 18]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+        addLog(tr("Reading File.."));
+        while(!gcodestream.atEnd()){
+            cline = gcodestream.readLine();
+            cline = cline.simplified();
+            if(cline.contains(QChar(';'))){
+                cline.resize(cline.indexOf(QChar(';')));
+            }
+            if(!cline.isEmpty()){
+                pro->pushCommand(cline.toLocal8Bit());
+                bool waiting=true;
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: space

File: src/mainwindow.h (hunk 9 above)
Comment (sithlord48) [inline src/mainwindow.h]: yes . 
changed to "Extruder"

File: src/mainwindow.cpp [hunk 19]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+                while(waiting){
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: (rtnCommand != "ok 0" && rtnCommand != "wait")

File: src/mainwindow.cpp (hunk 10 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: will add in disconnect

File: src/mainwindow.cpp (hunk 14 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: removed.

File: src/mainwindow.cpp (hunk 15 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: removed.

File: src/mainwindow.cpp (hunk 16 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: added indent

File: src/mainwindow.cpp (hunk 17 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: this line is not ment to stay . will be removed.

File: src/mainwindow.cpp (hunk 13 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: removed

File: src/mainwindow.cpp (hunk 12 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: removed

File: src/mainwindow.cpp [hunk 20]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+            if(!cline.isEmpty()){
+                pro->pushCommand(cline.toLocal8Bit());
+                bool waiting = true;
+                while(waiting){
+                    if(!pro->commandAvailable()){
+                        loop.exec();
+                    }
+                    //Handle Messages from printer.
+                    if (rtnCommand == "ok 0"){
+                        waiting=false;
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: space

File: src/mainwindow.cpp [hunk 21]
```diff
@@ -102,31 +114,132 @@ void MainWindow::locateSerialPort()
...
+                    }
+
+                    else if (rtnCommand == "wait"){
+                        waiting=false;
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: space

Comment (sithlord48) [inline src/mainwindow.cpp]: spaced added

File: src/mainwindow.cpp (hunk 20 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: space added

**Comment by patrickelectric** [comment]
@sithlord48, Everything ok to move to protocollayer.

File: src/mainwindow.cpp [hunk 22]
```diff
@@ -206,39 +205,7 @@ void MainWindow::printPBClicked()
...
-
-                    else if (rtnCommand == "wait"){
-                        waiting = false;
-                    }
-                    /* Handle resend and ok 1 soon*/
-                }
-            }
-        }
+        pro->print(fileName);
         addLog(tr("Finished Reading File"));
```
Comment (patrickelectric) [inline src/mainwindow.cpp]: It's better to protocollayer emit a signal to perform the finished print.
Right now is a hardcoded function, it's better to do it now.

File: src/protocollayer.cpp [hunk 23]
```diff
@@ -1,10 +1,57 @@
 #include "protocollayer.h"
 
+#include <QFile>
+#include <QEventLoop>
+#include <QTextStream>
+
 ProtocolLayer::ProtocolLayer(QString port, uint baud, QWidget *parent) :
 	SerialLayer(port, baud, parent)
 {
+    lastMessage="";
```
Comment (patrickelectric) [inline src/protocollayer.cpp]: Start the variables like this,
```
ProtocolLayer::ProtocolLayer(QString port, uint baud, QWidget *parent) :
    SerialLayer(port, baud, parent),
    lastMessage(QByteArray())
```

File: src/protocollayer.cpp [hunk 24]
```diff
@@ -1,10 +1,57 @@
...
+}
+
+void ProtocolLayer::print(QString fileName)
+{
+    QFile file(fileName);
+    file.open(QFile::ReadOnly);
+    QTextStream gcodestream(&file);
+    QString cline;
+    QEventLoop loop;
+    connect(this, &ProtocolLayer::receivedCommand, &loop, &QEventLoop::quit);
```
Comment (patrickelectric) [inline src/protocollayer.cpp]: You need to disconnect in the end, check the [doc](http://doc.qt.io/qt-5/qobject.html).

File: src/mainwindow.h [hunk 25]
```diff
@@ -16,8 +16,9 @@ Q_OBJECT
 
 public slots:
     /**
-     * @brief Check received message
+     * @brief Check received messages
      *
+     * @param msg : Message
```
Comment (patrickelectric) [inline src/mainwindow.h]: Why this ? The function don't receive param, check [doc](https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html).

**Comment by patrickelectric** [comment]
I will wait @sithlord48 to solve the last comments, to make the comments about the commit order and organization (it will be _some_ comments).
@sithlord48, If you like organize your commits before it will be great :)
Try to separete in some commits like: Removing unnused variable in protocolllayer, Correcting docummentation, add new interface and other things like that.

File: src/protocollayer.cpp (hunk 24 above)
Comment (sithlord48) [inline src/protocollayer.cpp]: added disconnect after the loop.

File: src/protocollayer.cpp (hunk 23 above)
Comment (sithlord48) [inline src/protocollayer.cpp]: :heavy_check_mark: done

File: src/mainwindow.cpp (hunk 22 above)
Comment (sithlord48) [inline src/mainwindow.cpp]: This message is ment to be removed. 
A Generic info loging function need to be made within the Protocol layer to send this these kinds of messages. After that log messages like this and more will be added to the Protocol layer.

File: src/mainwindow.h (hunk 25 above)
Comment (sithlord48) [inline src/mainwindow.h]: :heavy_check_mark:  removed!

**Pull request merged**